sdk/src/azure/core/az_hfsm.c

```diff
@@ -0,0 +1,164 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file
+ * @brief Hierarchical Finite State Machine (HFSM) implementation.
+ *
+ * @details This implementation is _not_ providing complete HFSM functionality. The following
+ *          constraints must be made by the developer for their state machines:
+ *          1. A single top level state must exist.
+ *          2. Transitions can only be made to sub-states, peer-states and super-states.
+ *          3. The initial state is always the top-level state. Transitions must be made by the
+ *             application if an inner state must be reached during initialization.
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <azure/core/internal/az_hfsm_internal.h>
+#include <azure/core/internal/az_precondition_internal.h>
+#include <azure/core/internal/az_result_internal.h>
+
+#include <azure/core/_az_cfg.h>
+
+const az_event _az_hfsm_event_entry = { AZ_HFSM_EVENT_ENTRY, NULL };
+const az_event _az_hfsm_event_exit = { AZ_HFSM_EVENT_EXIT, NULL };
+
+static az_result _az_hfsm_event_handler(az_event_policy* h, az_event event)
+{
+  return _az_hfsm_send_event((_az_hfsm*)h, event);
+}
+
+AZ_NODISCARD az_result _az_hfsm_init(
+    _az_hfsm* h,
+    az_event_policy_handler root_state,
+    _az_hfsm_get_parent get_parent_func,
+    az_event_policy* outbound_policy,
+    az_event_policy* inbound_policy)
+{
+  _az_PRECONDITION_NOT_NULL(h);
+  _az_PRECONDITION_NOT_NULL(root_state);
+  _az_PRECONDITION_NOT_NULL(get_parent_func);
+  h->_internal.current_state = root_state;
+  h->_internal.get_parent_func = get_parent_func;
+
+  h->policy.inbound_handler = _az_hfsm_event_handler;
+  h->policy.outbound_handler = _az_hfsm_event_handler;
+
+  h->policy.outbound_policy = outbound_policy;
+  h->policy.inbound_policy = inbound_policy;
+
+  az_result ret = h->_internal.current_state((az_event_policy*)h, _az_hfsm_event_entry);
+
+  // The HFSM should never return this result.
+  _az_PRECONDITION(ret != AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE);
+  return ret;
+}
+
+static AZ_NODISCARD az_result
+__az_hfsm_recursive_exit(_az_hfsm* h, az_event_policy_handler source_state)
```

Comment:
A single `_` only please (search for others that we might have missed).
Also, ordering:

`AZ_NODISCARD static az_result` (everywhere).